helix-core/src/main/java/org/apache/helix/controller/stages/task/TaskSchedulingStage.java

```diff
@@ -81,13 +81,16 @@ public void process(ClusterEvent event) throws Exception {
     // Reset current INIT/RUNNING tasks on participants for throttling
     cache.resetActiveTaskCount(currentStateOutput);
 
+    ClusterStatusMonitor clusterStatusMonitor =
+        event.getAttribute(AttributeName.clusterStatusMonitor.name());
     buildQuotaBasedWorkflowPQsAndInitDispatchers(cache,
-        (HelixManager) event.getAttribute(AttributeName.helixmanager.name()),
-        (ClusterStatusMonitor) event.getAttribute(AttributeName.clusterStatusMonitor.name()));
+        (HelixManager) event.getAttribute(AttributeName.helixmanager.name()), clusterStatusMonitor);
 
     final BestPossibleStateOutput bestPossibleStateOutput =
         compute(event, resourceMap, currentStateOutput);
     event.addAttribute(AttributeName.BEST_POSSIBLE_STATE.name(), bestPossibleStateOutput);
+
+    cache.getAssignableInstanceManager().recordAvailableThreadsPerType(clusterStatusMonitor);
```

Comment:
Have we done this kind of update before? I remember we usually have the method in clusterStatusMonitor to have the metrics logic in that place. If we do like this, then we are distributing monitoring logic in different places. Could be a potential issue.

Reply:
The reason why I did it this way is because `AssignableInstanceManager` already keeps a map of available threads per job type across the cluster, so if we have another place keeping track of the threads, it seems to be duplication. 
I want to draw parallel to how `AbstractTaskDispatcher` uses cluster status monitor: we set the monitor as its local variable, then the dispatcher uses the monitor. It's pretty similar to this use case in my opinion.

Reply:
Where to keep the data should be fine. But it depends on the where the logic is. 
You can do the reserve way to update it. For example, you can pass in the cache AssignableInstanceManager as an argument for clusterStatusMonitor.updateQuota() (you can name it differently). But keep the logic of how you update the data inside the clusterStatusMonitor. You dont have to copy the data to clusterStatusMonitor.

Reply:
For the sake of cutting down on redundant computation, is it possible at all to move this "record" (basically a few loops that count) logic to where the global quota calculation is done so we don't loop through things twice?

Reply:
@junkaixue I see, let me try that. 
@narendly I considered that earlier: the global quota is updated in several methods; it's not just one computation at a central place. Plus, based on what @junkaixue said, we want to "record" inside `ClusterStatusMonitor`, instead of doing it in `AssiganbleInstanceManager`. Lastly, it's 1 loop through each job type, which I don't think is too computation heavy (it doesn't loop through all instances).